# Разведочный анализ данных по продаже автомобилей

Цель работы - изучить датасет с объявлениями о продаже автомобилей, выполнить очистку данных и подготовить итоговый датасет для следующих лабораторных работ по построению модели прогнозирования цены продажи.


## Загрузка данных и знакомство с ними

In [5]:
from pathlib import Path
from datetime import datetime

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

sns.set_theme(style="whitegrid")
plt.rcParams["figure.figsize"] = (10, 6)

cwd = Path.cwd().resolve()
if (cwd / "data").exists() and (cwd / "eda").exists():
    ROOT_DIR = cwd
elif (cwd.parent / "data").exists() and (cwd.parent / "eda").exists():
    ROOT_DIR = cwd.parent
else:
    raise FileNotFoundError("Не удалось определить корневую директорию проекта.")

DATA_DIR = ROOT_DIR / "data"
EDA_DIR = ROOT_DIR / "eda"
RAW_DATA_PATH = DATA_DIR / "car_data.csv"
CLEAN_CSV_PATH = DATA_DIR / "car_data_cleaned.csv"
CLEAN_PKL_PATH = DATA_DIR / "car_data_cleaned.pkl"

raw_df = pd.read_csv(RAW_DATA_PATH)
df = raw_df.copy()
target_column = "Selling_Price"

print(f"Исходный датасет: {RAW_DATA_PATH}")
print(f"Размер исходного датасета: {df.shape}")
print(f"Целевая переменная: {target_column}")

Исходный датасет: C:\Users\mishu\Mohov\lab1\data\car_data.csv
Размер исходного датасета: (301, 9)
Целевая переменная: Selling_Price


In [6]:
print("Типы данных по столбцам:")
print(df.dtypes.to_string())

numeric_cols_initial = df.select_dtypes(include=[np.number]).columns.tolist()
categorical_cols_initial = df.select_dtypes(include=["object", "string", "category"]).columns.tolist()

print()
print("Числовые признаки:")
print(numeric_cols_initial)

print()
print("Категориальные признаки:")
print(categorical_cols_initial)

print()
print("Основные статистики для числовых признаков:")
print(df[numeric_cols_initial].describe().round(2).to_string())

categorical_summary = pd.DataFrame(
    {
        "Количество уникальных значений": df[categorical_cols_initial].nunique(),
        "Примеры значений": [
            ", ".join(map(str, df[col].astype(str).unique()[:5]))
            for col in categorical_cols_initial
        ],
    }
)

print()
print("Краткая сводка по категориальным признакам:")
print(categorical_summary.to_string())

Типы данных по столбцам:
Car_Name             str
Year               int64
Selling_Price    float64
Present_Price    float64
Driven_kms         int64
Fuel_Type            str
Selling_type         str
Transmission         str
Owner              int64

Числовые признаки:
['Year', 'Selling_Price', 'Present_Price', 'Driven_kms', 'Owner']

Категориальные признаки:
['Car_Name', 'Fuel_Type', 'Selling_type', 'Transmission']

Основные статистики для числовых признаков:
          Year  Selling_Price  Present_Price  Driven_kms   Owner
count   301.00         301.00         301.00      301.00  301.00
mean   2013.63           4.66           7.63    36947.21    0.04
std       2.89           5.08           8.64    38886.88    0.25
min    2003.00           0.10           0.32      500.00    0.00
25%    2012.00           0.90           1.20    15000.00    0.00
50%    2014.00           3.60           6.40    32000.00    0.00
75%    2016.00           6.00           9.90    48767.00    0.00
max    2018.00 

**Вывод по разделу.** В исходном датасете 301 запись и 9 признаков. Целевая переменная - `Selling_Price`. До очистки числовыми являются `Year`, `Selling_Price`, `Present_Price`, `Driven_kms`, `Owner`, а категориальными - `Car_Name`, `Fuel_Type`, `Selling_type`, `Transmission`. Признак `Owner` логично рассматривать как категориальный, так как он описывает количество владельцев, а не непрерывную величину.

## Очистка данных

In [7]:
cleaning_actions = []

df_clean = df.copy()

duplicates_count = int(df_clean.duplicated().sum())
df_clean = df_clean.drop_duplicates().copy()
cleaning_actions.append(f"Удалено полных дубликатов: {duplicates_count}.")

current_year = datetime.now().year
invalid_mask = (
    (df_clean["Year"] < 1980)
    | (df_clean["Year"] > current_year)
    | (df_clean["Selling_Price"] <= 0)
    | (df_clean["Present_Price"] <= 0)
    | (df_clean["Driven_kms"] < 0)
)
invalid_rows = int(invalid_mask.sum())
if invalid_rows:
    df_clean = df_clean.loc[~invalid_mask].copy()
    cleaning_actions.append(f"Удалено невалидных строк по базовым правилам: {invalid_rows}.")
else:
    cleaning_actions.append("Невалидных строк по базовым правилам не обнаружено.")

missing_total = int(df_clean.isna().sum().sum())
if missing_total:
    numeric_for_fill = df_clean.select_dtypes(include=[np.number]).columns.tolist()
    categorical_for_fill = df_clean.select_dtypes(include=["object", "string", "category"]).columns.tolist()

    for col in numeric_for_fill:
        if df_clean[col].isna().any():
            df_clean[col] = df_clean[col].fillna(df_clean[col].median())

    for col in categorical_for_fill:
        if df_clean[col].isna().any():
            df_clean[col] = df_clean[col].fillna(df_clean[col].mode().iloc[0])

    cleaning_actions.append("Пропущенные значения восстановлены: числовые медианой, категориальные модой.")
else:
    cleaning_actions.append("Пропущенных значений не обнаружено, восстановление не потребовалось.")

category_cols = ["Car_Name", "Fuel_Type", "Selling_type", "Transmission", "Owner"]
for col in category_cols:
    df_clean[col] = df_clean[col].astype("category")

cleaning_actions.append("Категориальные признаки приведены к типу category, включая Owner.")

numeric_feature_cols = [col for col in df_clean.select_dtypes(include=[np.number]).columns if col != target_column]
categorical_feature_cols = df_clean.select_dtypes(include=["object", "string", "category"]).columns.tolist()

print("Размер до очистки:", raw_df.shape)
print("Размер после очистки:", df_clean.shape)
print()
print("Действия очистки:")
for action in cleaning_actions:
    print("-", action)

print()
print("Пропуски после очистки:")
print(df_clean.isna().sum().to_string())

print()
print("Типы данных после очистки:")
print(df_clean.dtypes.to_string())

Размер до очистки: (301, 9)
Размер после очистки: (299, 9)

Действия очистки:
- Удалено полных дубликатов: 2.
- Невалидных строк по базовым правилам не обнаружено.
- Пропущенных значений не обнаружено, восстановление не потребовалось.
- Категориальные признаки приведены к типу category, включая Owner.

Пропуски после очистки:
Car_Name         0
Year             0
Selling_Price    0
Present_Price    0
Driven_kms       0
Fuel_Type        0
Selling_type     0
Transmission     0
Owner            0

Типы данных после очистки:
Car_Name         category
Year                int64
Selling_Price     float64
Present_Price     float64
Driven_kms          int64
Fuel_Type        category
Selling_type     category
Transmission     category
Owner            category


**Вывод по разделу.** На этапе очистки были удалены дубликаты, проверены базовые правила валидности для года выпуска, цены и пробега, а также скорректированы типы данных. Пропущенных значений в датасете не было, поэтому дополнительное восстановление не потребовалось.

## Анализ признаков для модели

In [ ]:
fig = px.histogram(
    df_clean,
    x=target_column,
    nbins=40,
    title="Интерактивное распределение цены продажи",
    labels={target_column: "Цена продажи"},
)
fig.update_layout(bargap=0.08)
fig.write_html(EDA_DIR / "selling_price_hist_interactive.html", include_plotlyjs="cdn")
fig.show()

print(df_clean[target_column].describe().round(2).to_string())

**Вывод по графику 1.** Распределение `Selling_Price` имеет выраженную правостороннюю асимметрию: основная масса автомобилей сосредоточена в нижнем ценовом диапазоне, а дорогие автомобили образуют длинный правый хвост. Это важно учитывать при выборе модели и при оценке влияния выбросов на целевую переменную.

In [ ]:
plt.figure(figsize=(9, 6))
sns.scatterplot(
    data=df_clean,
    x="Present_Price",
    y=target_column,
    hue="Fuel_Type",
    alpha=0.75,
)
plt.title("Связь текущей и продажной цены автомобиля")
plt.xlabel("Текущая цена автомобиля")
plt.ylabel("Цена продажи")
plt.tight_layout()
plt.savefig(EDA_DIR / "present_vs_selling_price.png", dpi=300, bbox_inches="tight")
plt.show()

corr_present = df_clean[["Present_Price", target_column]].corr().iloc[0, 1]
print(f"Корреляция Present_Price и Selling_Price: {corr_present:.3f}")

**Вывод по графику 2.** Между `Present_Price` и `Selling_Price` наблюдается сильная положительная зависимость: чем выше текущая цена автомобиля, тем выше цена продажи.

In [ ]:
fuel_order = (
    df_clean.groupby("Fuel_Type", observed=False)[target_column]
    .median()
    .sort_values(ascending=False)
    .index
)

plt.figure(figsize=(8, 6))
sns.boxplot(data=df_clean, x="Fuel_Type", y=target_column, order=fuel_order)
plt.title("Цена продажи в зависимости от типа топлива")
plt.xlabel("Тип топлива")
plt.ylabel("Цена продажи")
plt.tight_layout()
plt.savefig(EDA_DIR / "price_by_fuel_type.png", dpi=300, bbox_inches="tight")
plt.show()

print(df_clean.groupby("Fuel_Type", observed=False)[target_column].agg(["mean", "median", "count"]).round(2).to_string())

**Вывод по графику 3.** В текущем датасете автомобили с дизельным топливом в среднем продаются дороже бензиновых. Для CNG наблюдений слишком мало, чтобы делать уверенный вывод, но в целом тип топлива выглядит полезным категориальным признаком для модели.

In [ ]:
plt.figure(figsize=(8, 6))
sns.boxplot(data=df_clean, x="Transmission", y=target_column)
plt.title("Цена продажи в зависимости от типа трансмиссии")
plt.xlabel("Тип трансмиссии")
plt.ylabel("Цена продажи")
plt.tight_layout()
plt.savefig(EDA_DIR / "price_by_transmission.png", dpi=300, bbox_inches="tight")
plt.show()

print(df_clean.groupby("Transmission", observed=False)[target_column].agg(["mean", "median", "count"]).round(2).to_string())

**Вывод по графику 4.** У автомобилей с автоматической коробкой передач цена продажи заметно выше, чем у автомобилей с механической коробкой. Признак `Transmission` потенциально полезен для модели и может объяснять часть ценового разброса.

In [ ]:
top_cars = df_clean["Car_Name"].value_counts().head(15)

plt.figure(figsize=(10, 7))
plt.barh(top_cars.index.astype(str), top_cars.values, color="steelblue")
plt.title("Топ-15 самых часто встречающихся моделей")
plt.xlabel("Количество автомобилей")
plt.ylabel("Модель автомобиля")
plt.gca().invert_yaxis()
plt.tight_layout()
plt.savefig(EDA_DIR / "top_car_models.png", dpi=300, bbox_inches="tight")
plt.show()

print(top_cars.to_string())

**Вывод по графику 5.** В датасете сильнее всего представлены несколько популярных моделей, например `city`, `corolla altis` и `verna`. Это означает, что распределение по моделям неравномерно, и модель может быть чувствительна к более часто встречающимся классам автомобилей.

In [ ]:
owner_summary = df_clean.groupby("Owner", observed=False)[target_column].agg(["mean", "median", "count"]).round(2)

plt.figure(figsize=(8, 6))
sns.barplot(x=owner_summary.index.astype(str), y=owner_summary["mean"], color="indianred")
plt.title("Средняя цена продажи по количеству владельцев")
plt.xlabel("Количество владельцев")
plt.ylabel("Средняя цена продажи")
plt.tight_layout()
plt.savefig(EDA_DIR / "price_by_owner.png", dpi=300, bbox_inches="tight")
plt.show()

print(owner_summary.to_string())

**Вывод по графику 6.** Основная часть наблюдений относится к автомобилям без предыдущих владельцев, и именно по этой группе доступны наиболее устойчивые оценки. Для групп с 1 и особенно 3 владельцами наблюдений мало, но признак `Owner` всё равно стоит учитывать как категориальный, так как он отражает историю владения, а не количественную шкалу.

## Сохранение финального датасета

In [ ]:
df_clean.to_csv(CLEAN_CSV_PATH, index=False)
df_clean.to_pickle(CLEAN_PKL_PATH)

print(f"Очищенный датасет сохранен в CSV: {CLEAN_CSV_PATH}")
print(f"Очищенный датасет сохранен в pickle: {CLEAN_PKL_PATH}")
print()
print("Финальные типы данных:")
print(df_clean.dtypes.to_string())

## Выводы

In [24]:
correlations = df_clean[[*numeric_feature_cols, target_column]].corr(numeric_only=True)[target_column].sort_values(ascending=False)
key_correlations = correlations.drop(target_column)

model_insights = [
    "Распределение цены продажи имеет правый хвост, поэтому дорогие автомобили формируют выбросы по целевой переменной.",
    f"Наиболее сильная линейная связь с ценой наблюдается у Present_Price ({key_correlations['Present_Price']:.3f}).",
    f"Год выпуска положительно связан с ценой ({key_correlations['Year']:.3f}), то есть более новые автомобили обычно дороже.",
    "Тип топлива и тип трансмиссии заметно влияют на цену продажи.",
    "Количество владельцев можно использовать как категориальный признак, поскольку автомобили без предыдущих владельцев обычно стоят дороже.",
]

print("Действия, выполненные при очистке данных:")
for action in cleaning_actions:
    print("-", action)


print()
print("Закономерности, полезные для моделирования:")
for insight in model_insights:
    print("-", insight)

Действия, выполненные при очистке данных:
- Удалено полных дубликатов: 2.
- Невалидных строк по базовым правилам не обнаружено.
- Пропущенных значений не обнаружено, восстановление не потребовалось.
- Категориальные признаки приведены к типу category, включая Owner.

Закономерности, полезные для моделирования:
- Распределение цены продажи имеет правый хвост, поэтому дорогие автомобили формируют выбросы по целевой переменной.
- Наиболее сильная линейная связь с ценой наблюдается у Present_Price (0.876).
- Год выпуска положительно связан с ценой (0.234), то есть более новые автомобили обычно дороже.
- Тип топлива и тип трансмиссии заметно влияют на цену продажи.
- Количество владельцев можно использовать как категориальный признак, поскольку автомобили без предыдущих владельцев обычно стоят дороже.
